quickwit-search/src/root.rs

```diff
@@ -0,0 +1,353 @@
+            search_request: Some(search_request_with_offset_0.clone()),
+            split_uris: jobs.iter().map(|job| job.split.clone()).collect(),
+        };
+
+        let handle = tokio::spawn(async move {
+            match search_client.leaf_search(leaf_search_request).await {
+                Ok(resp) => Ok(resp.into_inner()),
+                Err(err) => Err(anyhow::anyhow!(
+                    "Failed to search a leaf node due to {:?}",
+                    err
+                )),
+            }
+        });
+        leaf_search_handles.push(handle);
+    }
+    let leaf_search_responses = futures::future::try_join_all(leaf_search_handles).await?;
+
+    // Find the sum of the number of hits and merge multiple partial hits into a single partial hits.
+    let mut num_hits = 0;
+    let mut partial_hits = Vec::new();
+    for response in leaf_search_responses {
+        match response {
+            Ok(leaf_search_result) => {
+                num_hits += leaf_search_result.num_hits;
+                partial_hits = merge_partial_hits(partial_hits, leaf_search_result.partial_hits);
+            }
+            Err(err) => error!(err=?err),
+        }
+    }
+    // Extract the leaf search results for the range specified in the original search request.
+    let start_offset = if search_request.start_offset > partial_hits.len() as u64 {
+        partial_hits.len() as u64
+    } else {
+        search_request.start_offset
+    };
+    let end_offset = if start_offset + search_request.max_hits > partial_hits.len() as u64 {
+        partial_hits.len() as u64
+    } else {
+        start_offset + search_request.max_hits
+    };
+    partial_hits = partial_hits
+        .as_slice()
+        .get(start_offset as usize..end_offset as usize)
+        .unwrap_or(&Vec::new())
+        .to_vec();
+    // Make the leaf search results after distributed search.
+    let leaf_search_result = LeafSearchResult {
+        num_hits,
+        partial_hits,
+    };
+
+    // Create a hash map of PartialHit with split as a key.
+    let mut partial_hits_map: HashMap<String, Vec<PartialHit>> = HashMap::new();
+    for partial_hit in leaf_search_result.partial_hits.iter() {
+        partial_hits_map
+            .entry(partial_hit.split.clone())
+            .or_insert_with(Vec::new)
+            .push(partial_hit.clone());
+    }
+
+    // Create a job for fetching docs and assign the splits that the node is responsible for based on the job.
+    let fetch_docs_jobs: Vec<Job> = partial_hits_map
+        .keys()
+        .map(|split| {
+            // TODO: Change to a better way that does not use unwrap().
+            let split_metadata = split_metadata_map.get(split).unwrap();
+            Job {
+                split: split.clone(),
+                cost: compute_split_cost(split_metadata),
+            }
+        })
+        .collect();
+    let assigned_fetch_docs_jobs = client_pool.assign_jobs(fetch_docs_jobs).await?;
```

Comment:
We do not want to reassign the jobs. We want to use the same node as what we used for the leaf code for this split.

Reply:
Changed.
Please see https://github.com/quickwit-inc/quickwit/pull/163/commits/649bd3e10860ddd82a7bb125ed494614f858e0c0.